pkg/discovery/systemd.go

```diff
@@ -0,0 +1,165 @@
+// Copyright 2021 The Parca Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package discovery
+
+import (
+	"bufio"
+	"context"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"sync"
+	"syscall"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/prometheus/common/model"
+)
+
+type SystemdConfig struct {
+	units map[string]struct{}
+}
+
+type SystemdDiscoverer struct {
+	units         map[string]struct{}
+	unitProfilers map[string]struct{}
+	logger        log.Logger
+
+	mtx sync.RWMutex
+}
+
+func (c *SystemdConfig) Name() string {
+	return "systemd"
+}
+
+func NewSystemdConfig(systemdUnits []string) *SystemdConfig {
+
+	units := map[string]struct{}{}
+
+	for _, unit := range systemdUnits {
+		units[unit] = struct{}{}
+	}
+	return &SystemdConfig{
+		units: units,
+	}
+}
+
+func (c *SystemdConfig) NewDiscoverer(d DiscovererOptions) (Discoverer, error) {
+	return &SystemdDiscoverer{
+		units:         c.units,
+		unitProfilers: make(map[string]struct{}),
+		logger:        d.Logger,
+	}, nil
+}
+
+func (c *SystemdDiscoverer) Run(ctx context.Context, up chan<- []*Group) error {
+	ticker := time.NewTicker(time.Second)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-ticker.C:
+		}
+		level.Debug(c.logger).Log("msg", "running systemd manager", "units", len(c.units))
+		var targetGroups []*Group
+
+		for unit := range c.units {
+
+			labelset, err := c.ReconcileUnit(ctx, unit)
+			if err != nil {
+				return err
+			}
+
+			labelset["systemd_unit"] = model.LabelValue(unit)
+
+			targetGroups = append(targetGroups, &Group{
+				Targets: []model.LabelSet{labelset},
+				Source:  unit,
+			})
+		}
+
+		up <- targetGroups
+	}
+}
+
+func (c *SystemdDiscoverer) ReconcileUnit(ctx context.Context, unit string) (model.LabelSet, error) {
+	f, err := os.Open(fmt.Sprintf("/sys/fs/cgroup/systemd/system.slice/%s/cgroup.procs", unit))
```

Comment:
These paths will conflict quite a bit with what is done in https://github.com/parca-dev/parca-agent/pull/171

I don't think we should wait on either front for the other one, just something to keep in mind, but wouldn't hurt to move the templating to a separate function I think that will probably make things easier either way.